tools/spanner/purge_ttl.py

```diff
@@ -23,57 +25,119 @@
 client = spanner.Client()
 
 
-def from_env():
+def use_dsn(args):
     try:
-        url = os.environ.get("SYNC_DATABASE_URL")
-        if not url:
+        if not args.sync_database_url:
             raise Exception("no url")
+        url = args.sync_database_url
         purl = parse.urlparse(url)
         if purl.scheme == "spanner":
             path = purl.path.split("/")
-            instance_id = path[-3]
-            database_id = path[-1]
+            args.instance_id = path[-3]
+            args.database_id = path[-1]
     except Exception as e:
         # Change these to reflect your Spanner instance install
         print("Exception {}".format(e))
-        instance_id = os.environ.get("INSTANCE_ID", "spanner-test")
-        database_id = os.environ.get("DATABASE_ID", "sync_stage")
-    return (instance_id, database_id)
+    return args
 
 
-def spanner_read_data(request=None):
-    (instance_id, database_id) = from_env()
-    instance = client.instance(instance_id)
-    database = instance.database(database_id)
+def deleter(database, name, query):
+    with statsd.timer("syncstorage.purge_ttl.{}_duration".format(name)):
+        logging.info("Running: {}".format(query))
+        start = datetime.now()
+        result = database.execute_partitioned_dml(query)
+        end = datetime.now()
+        logging.info(
+            "{name}: removed {result} rows, {name}_duration: {time}".format(
+            name=name, result=result, time=end - start))
+
+
+def add_conditions(args, query):
+    if args.collection_ids:
+        query += " AND collection_id"
+        if len(args.collection_ids) == 1:
+            query += " = {:d}".format(args.collection_ids[0])
+        else:
+            query += " in ({})".format(
+                ', '.join(map(str, args.collection_ids)))
+    return query
+
 
-    logging.info("For {}:{}".format(instance_id, database_id))
+def spanner_purge(args, request=None):
+    instance = client.instance(args.instance_id)
+    database = instance.database(args.database_id)
+
+    logging.info("For {}:{}".format(args.instance_id, args.database_id))
+    batch_query = add_conditions(
```

Comment:
Oh sorry, almost forgot: let's not apply this to batches.

They hopefully won't ever have timeout problems, their expiry index is sane: a "global" index or whatever spanner calls it, not interleaved in `user_collections`. They're always going to have much less expired data to purge and their expiry logic is also a little different: *all* batches expire (and rather quickly) regardless of collection_id.

The batch delete job currently takes less than 10 min on prod.